log/loki.go

```diff
@@ -0,0 +1,483 @@
+			oldLogs = oldLogs[:len(oldLogs)+oldCount]
+
+			t := time.Now()
+			cutOffIndex := sortAndSplitMsgs(oldLogs, cutOff)
+			if cutOffIndex == 0 {
+				continue
+			}
+			t1 := time.Since(t)
+
+			strms := h.createPushMessage(oldLogs, cutOffIndex, oldDropped)
+			if cutOffIndex > len(oldLogs) {
+				oldLogs = oldLogs[:0]
+				continue
+			}
+			oldLogs = oldLogs[:copy(oldLogs, oldLogs[cutOffIndex:])]
+			t2 := time.Since(t) - t1
+
+			var b bytes.Buffer
+			_, err := strms.WriteTo(&b)
+			if err != nil {
+				fmt.Printf("Error while marshaling logs for loki %s\n", err)
+				continue
+			}
+			size := b.Len()
+			t3 := time.Since(t) - t2 - t1
+
+			err = h.push(b)
+			if err != nil {
+				fmt.Printf("Error while sending logs to loki %s\n", err)
+				continue
+			}
+			t4 := time.Since(t) - t3 - t2 - t1
+
+			if h.profile {
+				fmt.Printf("sorting=%s, adding=%s marshalling=%s sending=%s count=%d final_size=%d\n",
+					t1, t2, t3, t4, cutOffIndex, size)
+			}
+		}
+	}()
+
+	for {
+		select {
+		case entry := <-h.ch:
+			if count == h.limit {
+				dropped++
+				continue
+			}
+
+			// Arguably we can directly generate the final marshalled version of the labels right here
+			// through sorting the entry.Data, removing additionalparams from it and then dumping it
+			// as the final marshal and appending level and h.labels after it.
+			// If we reuse some kind of big enough `[]byte` buffer we can also possibly skip on some
+			// of allocation. Combined with the cutoff part and directly pushing in the final data
+			// type this can be really a lot faster and to use a lot less memory
+			labels := make(map[string]string, len(entry.Data)+1)
+			for k, v := range entry.Data {
+				labels[k] = fmt.Sprint(v) // TODO optimize ?
+			}
+			for _, params := range h.labels {
+				labels[params[0]] = params[1]
+			}
+			labels["level"] = entry.Level.String()
+			// have the cutoff here ?
+			// if we cutoff here we can cut somewhat on the backbuffers and optimize the inserting
+			// in/creating of the final Streams that we push
+			msgs[count] = tmpMsg{
+				labels: labels,
+				msg:    entry.Message,
+				t:      entry.Time.UnixNano(),
+			}
+			count++
+		case t := <-ticker.C:
+			ch := make(chan int64)
+			pushCh <- ch
+			ch <- t.Add(-(h.pushPeriod / 2)).UnixNano()
+			<-ch
+		case <-h.ctx.Done():
+			ch := make(chan int64)
+			pushCh <- ch
+			ch <- 0
+			<-ch
+			return
+		}
+	}
+}
+
+func sortAndSplitMsgs(msgs []tmpMsg, cutOff int64) int {
+	if len(msgs) == 0 {
+		return 0
+	}
+
+	// TODO using time.Before was giving a lot of out of order, but even now, there are some, if the
+	// limit is big enough ...
+	sort.Slice(msgs, func(i, j int) bool {
+		return msgs[i].t < msgs[j].t
+	})
+
+	cutOffIndex := sort.Search(len(msgs), func(i int) bool {
+		return !(msgs[i].t < cutOff)
+	})
+	return cutOffIndex
+}
+
+func (h *lokiHook) createPushMessage(msgs []tmpMsg, cutOffIndex, dropped int) *lokiPushMessage {
+	strms := new(lokiPushMessage)
```

Comment:
Not sure if this was a typo or you tried to save 1 letter... If it's the second, please dont :pray:  :sweat_smile:

Reply:
:pray:

Reply:
are you talking about it being `strms` instead of `streams`? because that are 2 letters ;P

Reply:
ok, I apparently can't count or see, but the point still stands :sweat_smile: